### PR TITLE
feat: new router

### LINKED ISSUE #618: Port Angular.JS v1.4 router to AtScript / Angular2

### PR DESCRIPTION
**Pull request by btford**
~~~This is still WIP and pends on #1261, which needs to be rebased on top of @tbosch's changes that landed last night.~~~

~~~This is pretty much good to go. Needs some cleanup and review, but otherwise should be pretty okay.~~~

This is super ready to go! Let's merge!

Closes #618

### CODE REVIEW

File: modules/angular2/router.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/**
+ * @module
+ * @public
+ * @description
+ * Map URLs to Angular components
```
Comment (mhevery) [inline modules/angular2/router.js]: Needs better description. Maybe:

Maps application URLs into application states, to support deep-linking and navigation.

File: modules/angular2/src/facade/lang.dart [hunk 2]
```diff
@@ -25,6 +25,7 @@ bool isPresent(obj) => obj != null;
 bool isBlank(obj) => obj == null;
 bool isString(obj) => obj is String;
 bool isFunction(obj) => obj is Function;
+bool isClass(obj) => obj is Type;
```
Comment (mhevery) [inline modules/angular2/src/facade/lang.dart]: Would `isType` be more descriptive? We use the `Type` terminology everywhere else, and I think it better aligns with both Dart and ES5.

File: modules/angular2/src/router/instruction.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
+import {Map, MapWrapper, List, ListWrapper} from 'angular2/src/facade/collection';
+import {Promise, PromiseWrapper} from 'angular2/src/facade/async';
+import {isPresent} from 'angular2/src/facade/lang';
+
+export class RouteParams {
```
Comment (mhevery) [inline modules/angular2/src/router/instruction.js]: I think all of these need docs. (I am OK, with doing docs in separate PR)

File: modules/angular2/src/router/outlet_ref.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+  constructor(viewContainer, compiler:Compiler, injector:Injector) {
+    this.viewContainer = viewContainer;
+    this.compiler = compiler;
+    this.injector = injector;
+  }
+
+  activate(instruction:Instruction) {
+    return this.compiler.compileInHost(instruction.component).then((pv) => {
+      var inj = this.injector.resolveAndCreateChild([
```
Comment (mhevery) [inline modules/angular2/src/router/outlet_ref.js]: `inj` => `outletInjector`

File: modules/angular2/test/router/outlet_spec.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+})
+class HelloCmp {
+  greeting:string;
+  constructor() {
+    this.greeting = "hello";
+  }
+}
+
+
+@View({
```
Comment (mhevery) [inline modules/angular2/test/router/outlet_spec.js]: `@View` without `@Component` does not look right. Is something missing?

File: modules/angular2/src/router/outlet_ref.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import {Injector, bind} from 'angular2/di';
+import {Promise, PromiseWrapper} from 'angular2/src/facade/async';
+import {Instruction, RouteParams} from './instruction'
+import {DynamicComponentLoader} from 'angular2/src/core/compiler/dynamic_component_loader';
+import {ViewContainer} from 'angular2/src/core/compiler/view_container';
+import {isPresent} from 'angular2/src/facade/lang';
+import {ElementRef} from 'angular2/src/core/compiler/element_injector';
+import {Compiler} from 'angular2/src/core/compiler/compiler';
+
+export class OutletRef {
```
Comment (mhevery) [inline modules/angular2/src/router/outlet_ref.js]: Why is `OutletRef` and `RouterOutlet` two separate classes? Seems like one is just a facade for the other. Can we just have one class, or are there specific reasons?

Comment (mhevery) [inline modules/angular2/src/router/outlet_ref.js]: It looks like `OutletRef` implements some sort of interface. Did you forget `@Implements`?

File: modules/angular2/src/router/router.js [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    }
+    return mapObjAsync(instruction.viewports, fn)
+      .then(() => mapObjAsync(instruction.viewports, (childInstruction, _) => {
+        return childInstruction.router._traverseInstruction(childInstruction, fn);
+      }));
+  }
+
+
+  _activateOutlets(instruction:Instruction):Promise {
+    return this._queryViewports((port, name) => {
```
Comment (mhevery) [inline modules/angular2/src/router/router.js]: `port` and `name` should have a type.

File: modules/angular2/src/router/outlet_ref.js (hunk 6 above)
Comment (mhevery) [inline modules/angular2/src/router/outlet_ref.js]: My guess is that `OutletRef` should be an interface. And `RouterOutlet` should implement the `OutletRef`. Also I think `OutletRef` the `Ref` part bothers me. Should the interface be called `RoutePort` or just `Route`, not sure.

File: modules/angular2/src/router/outlet_ref.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        bind(Router).toValue(instruction.router)
+      ]);
+
+      this.viewContainer.clear();
+      this.viewContainer.create(0, pv, inj);
+    });
+  }
+
+  canActivate(instruction:any) {
+    return PromiseWrapper.resolve(true);
```
Comment (mhevery) [inline modules/angular2/src/router/outlet_ref.js]: Why do we have to wrap in promise. Can`t the router just do instance of and auto resolve for us?

File: modules/angular2/src/router/router.js [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+import {OutletRef} from './outlet_ref';
+
+/**
+ * # Router
+ * The router is responsible for mapping URLs to components.
+ *
+ * You can see the state of the router by inspecting the read-only field `router.navigating`.
+ * This may be useful for showing a spinner, for instance.
+ *
+ * @exportedAs angular2/router
```
Comment (mhevery) [inline modules/angular2/src/router/router.js]: all public API should have `@exportedAs angular2/router` I think whole bunch of classes are public, have no docs and are not marked as public.

File: modules/angular2/test/router/outlet_spec.js [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  template: "one"
+})
+class OneCmp {}
+
+@View({
+  template: "two"
+})
+class TwoCmp {}
```
Comment (mhevery) [inline modules/angular2/test/router/outlet_spec.js]: not used delete.

**Comment by mhevery** [comment]
Nice work sir. I think there is some confusion about interfaces etc, see comments. Can you clean them up and resubmit it. I think we are pretty close. Very exciting!

**Comment by btford** [comment]
On it. Thanks for the feedback!

File: modules/angular2/src/facade/lang.dart (hunk 2 above)
Comment (btford) [inline modules/angular2/src/facade/lang.dart]: :+1:

File: modules/angular2/src/router/outlet_ref.js (hunk 4 above)
Comment (btford) [inline modules/angular2/src/router/outlet_ref.js]: :+1:

File: modules/angular2/src/router/router.js (hunk 7 above)
Comment (btford) [inline modules/angular2/src/router/router.js]: Correct. Renamed accordingly.

Should anonymous functions always have types for their params?

File: modules/angular2/test/router/outlet_spec.js (hunk 10 above)
Comment (btford) [inline modules/angular2/test/router/outlet_spec.js]: :+1:

**Comment by btford** [comment]
I'm not sure why CI is failing – The failure seems like it comes from bower.

@mhevery @vsavkin @tbosch Any insights? I'd really like to get this merged ASAP.

**Comment by naomiblack** [comment]
known CI issue. @IgorMinar and @alexeagle are working on it.

**Comment by btford** [comment]
:disappointed:

**Comment by btford** [comment]
Landed as 1b2754dacdd15e8fea429d56cdacb28eae76d2b1 and e617ca6323902bd98c0f1eb990b82f6b8d3c98e3

:tada: :cake: 

Now I owe you some docs PRs.

**Comment by ocombe** [comment]
\o/
Now let's make those examples! :)

**Pull request closed without merging**